### PR TITLE
2FA setup during login

### LINKED ISSUE #12268: Possibility to set up 2FA on login
Originally discussed in https://github.com/nextcloud/server/issues/11021 and brought up in https://github.com/nextcloud/server/issues/12249, it would be a UX enhancement if we could allow users setting up their 2FA on login in case it was enforced by admins.

This will require some dedicated setup pages that are accessible even when 2FA is enforced, hence require a restructuring of how 2FA providers works and communicate with the server.

### PR DESCRIPTION
**Pull request by rullzer**
Hacky totp branch: https://github.com/nextcloud/twofactor_totp/pull/550

Fixes https://github.com/nextcloud/server/issues/12268

* [x] Make sure the setup during login routes are only available when 2FA is not yet setup.

### CODE REVIEW

File: core/Controller/TwoFactorChallengeController.php [hunk 1]
```diff
@@ -107,13 +108,15 @@ public function selectChallenge($redirect_url) {
...
 		list($providers, $backupProvider) = $this->splitProvidersAndBackupCodes($allProviders);
+		$setupProviders = $this->twoFactorManager->getLoginSetupProviders($user);
 
 		$data = [
 			'providers' => $providers,
 			'backupProvider' => $backupProvider,
 			'providerMissing' => $providerSet->isProviderMissing(),
 			'redirect_url' => $redirect_url,
 			'logout_url' => $this->getLogoutUrl(),
+			'setupProviers' => $setupProviders,
```
Comment (tcitworld) [inline core/Controller/TwoFactorChallengeController.php]: Typo :)

File: .htaccess [hunk 2]
```diff
@@ -65,3 +65,7 @@ Options -Indexes
 <IfModule pagespeed_module>
   ModPagespeed Off
 </IfModule>
+#### DO NOT CHANGE ANYTHING ABOVE THIS LINE ####
```
Comment (rullzer) [inline .htaccess]: whoops

**Comment by rullzer** [comment]
issue: If no setup providers are enabled I still get:

2fa

**Comment by rullzer** [comment]
issue: cancel login doesn't work when on setup page

**Comment by rullzer** [comment]
I tested the part from @ChristophWurst (thanks for finishing up!) good from my PoV.

**Comment by rullzer** [comment]
Ah. One more thing that came to mind. Should we have an admin setting if 2FA during login is allowed?

**Comment by ChristophWurst** [comment]
> Ah. One more thing that came to mind. Should we have an admin setting if 2FA during login is allowed?

Fine by me, though low priority. This can be added in a follow-up PR :wink:

**Comment by georgehrke** [comment]
336B494C-30D0-4E1D-BFBF-A057E578358E

GET http://nextcloud.local/apps/twofactor_totp/js/main-login-setup.js?v=1466e706-0 net::ERR_ABORTED 404 (Not Found)

@ChristophWurst Is this maybe related to your js cleanup?

**Comment by ChristophWurst** [comment]
> @ChristophWurst Is this maybe related to your js cleanup?

Were you on https://github.com/nextcloud/twofactor_totp/pull/550?

**Comment by georgehrke** [comment]
yes

**Comment by ChristophWurst** [comment]
> yes and no. Required Krankerl and there are no macOS builds afaik?

Yes, good point. I'm building it for OSX on CI but there are no binary releases (yet). Let me package that for you: [twofactor_totp.tar.gz](https://github.com/nextcloud/server/files/3186852/twofactor_totp.tar.gz)

**Comment by georgehrke** [comment]
> Yes, good point. I'm building it for OSX on CI but there are no binary releases (yet). 

That would be awesome. I was just about to simply wrap it in a docker container.

**Review by georgehrke — APPROVED** [review]
👍 Works

From a usability pov i would probably drop the second time, that you have to enter your code.

**Review by MorrisJobke — APPROVED** [review]
Tested and worked 👍

**Pull request merged**